### ISSUE TITLE
empty diagnostic method list from amm public facet etc.

### ISSUE THREAD

**Issue by dckc** [Agoric/Agoric#6570]
## Describe the bug

The "no method" diagnostic from the AMM public facet used to include the list of names of methods that it _does_ have. Now that list is empty.

## To Reproduce

In a REPL:

```
command[10] E(home.agoricNames).lookup('instance', 'amm')
history[10] [Object Alleged: InstanceHandle]{}
command[11] E(home.zoe).getPublicFacet(h[10])
history[11] [Object Alleged: AMM publicFacet]{}
command[12] E(h[11]).help()
history[12] Promise.reject("TypeError: target has no method \"help\", has []")
```

## Expected behavior

The `has []` list should include the available methods; for example:

```
command[13] E(home.zoe).getPublicFacet(E(home.agoricNames).lookup('instance', 'Pegasus'))
history[13] [Object Alleged: pegasus]{}
command[14] E(h[13]).help()
history[14] Promise.reject("TypeError: target has no method \"help\", has [\"getLocalIssuer\",\"makeInvitationToTransfer\",\"makePegasusConnectionKit\"]")
```

## Platform Environment
 - what OS are you using? what version of Node.js? v16.17.0
 - is there anything special/unusual about your platform?
 - what version of the Agoric-SDK are you using? (run `git describe --tags --always`)

@agoric/vat-data@0.4.3-209-g095c89a29

## Additional context

I suspect this is due to wrapping the AMM public facet in virtual / durable API stuff.

cc @michaelfig @Chris-Hibbert @turadg

**Comment by Chris-Hibbert** [comment]
The AMM's publicFacet is built with `augmentVirtualPublicFacet`, which is defined in `contractHelper.js` as 
```
  const augmentVirtualPublicFacet = originalPublicFacet => {
    return Far('publicFacet', {
      ...originalPublicFacet,
      ...commonPublicMethods,
      ...objectMap(typedAccessors, ignoreContext),
    });
  };
```
This hasn't changed recently, and I thought `Far` classes were more self-revealing than that.  Does that Alleged AMM publicFacet respond to the actual AMM public methods,  like `getLiquiditySupply`, `getQuoteIssuer`, or ` getAllPoolBrands`?

**Comment by dckc** [comment]
> ... Does that Alleged AMM publicFacet respond to the actual AMM public methods, like `getLiquiditySupply`, `getQuoteIssuer`, or ` getAllPoolBrands`?

yes:

```
command[15] E(h[11]).getAllPoolBrands()
history[15] [[Object Alleged: ATOM brand]{},[Object Alleged: BLD brand]{},[Object Alleged: LINK brand]{},[Object Alleged: USDC brand]{},[Object Alleged: WETH brand]{}]
```

`ignoreContext` seems sort of magical... I wonder if it introduces a conflict somehow.

**Comment by michaelfig** [comment]
The code that displays the diagnostic uses `Reflect.ownKeys(obj)` to fetch the properties.  Very likely with the introduction of Far Classes, method properties are moved somewhere else in the prototype chain (not as "own" properties of the object identity).

@erights, do you care to weigh in on this?

**Comment by Chris-Hibbert** [comment]
> `ignoreContext` seems sort of magical... I wonder if it introduces a conflict somehow.

ignoreContext is just taking functions of N arguments and dropping the first arg before calling them. It shouldn't have an impact like this.

**Comment by dckc** [comment]
search keyword: help

**Comment by dckc** [comment]
Good to see that land, but I think this needs an endo update before it's fixed.

**Comment by erights** [comment]
@dckc do you agree that this is fixed by https://github.com/endojs/endo/pull/1382 and can be closed?

I think it wasn't closed at the time only because it agoric-sdk until upgraded to depend on an endo post that PR, which happened long ago.

**Comment by dckc** [comment]
No; I just tried the REPL again, and the symptoms persist:

```
command[0] E(agoric.agoricNames).lookup('instance', 'amm')
history[0] [Object Alleged: InstanceHandle]{}
command[1] h=history,0
history[1] 0
command[2] ...
command[3] amm = {instance: h[0]}
history[3] {"instance":[Object Alleged: InstanceHandle]{}}
command[4] E(home.zoe).getPublicFacet(amm.instance)
history[4] [Object Alleged: AMM publicFacet]{}
command[5] amm.pub = h[4]
history[5] [Object Alleged: AMM publicFacet]{}
command[6] E(amm.pub).help()
history[6] Promise.reject("TypeError: target has no method \"help\", has []")
```

I don't see anything useful in the chain-side stack trace:

```
2023-01-30T17:07:43.223Z SwingSet: ls: v37: Logging sent error stack (TypeError#1)
2023-01-30T17:07:43.223Z SwingSet: ls: v37: TypeError#1: target has no method help , has []
2023-01-30T17:07:43.223Z SwingSet: ls: v37: TypeError: target has no method "help", has []
 at makeError (/bundled-source/.../node_modules/ses/src/error/assert.js:242)
 at fail (/bundled-source/.../node_modules/ses/src/error/assert.js:370)
 at localApplyMethod (/bundled-source/.../node_modules/@endo/eventual-send/src/local.js:116)
 at apply ()
 at dispatchToHandler (/bundled-source/.../node_modules/@endo/eventual-send/src/handled-promise.js:163)
 at win (/bundled-source/.../node_modules/@endo/eventual-send/src/handled-promise.js:488)
 at ()

2023-01-30T17:07:43.223Z SwingSet: ls: v37: TypeError#1 ERROR_NOTE: Sent as error:liveSlots:v37#70001
```

**Comment by dckc** [comment]
Odd... it works for some (`home.agoricNames`, ...) but not for others (`home.zoe`, ...):

```
command[7] E(home.zoe).help()
history[7] Promise.reject("TypeError: target has no method \"help\", has []")

command[8] E(home.agoricNames).help()
history[8] Promise.reject("TypeError: target has no method \"help\", has [\"entries\",\"keys\",\"lookup\",\"values\"]")
command[9] Object.keys(agoric)
history[9] ["agoricNames","attMaker","bank","board","chainTimerService","faucet","ibcport","myAddressNameAdmin","namesByAddress","priceAuthority","priceAuthorityAdminFacet","vaultFactoryCreatorFacet","zoe"]
command[10] E(home.priceAuthority).help()
history[10] Promise.reject("TypeError: target has no method \"help\", has [\"getQuoteIssuer\",\"getTimerService\",\"makeQuoteNotifier\",\"mutableQuoteWhenGT\",\"mutableQuoteWhenGTE\",\"mutableQuoteWhenLT\",\"mutableQuoteWhenLTE\",\"quoteAtTime\",\"quoteGiven\",\"quoteWanted\",\"quoteWhenGT\",\"quoteWhenGTE\",\"quoteWhenLT\",\"quoteWhenLTE\"]")
...
command[14] E(agoric.attMaker).help()
history[14] Promise.reject("TypeError: target has no method \"help\", has []")
command[15] E(agoric.board).help()
history[15] Promise.reject("TypeError: target has no method \"help\", has [\"getId\",\"getPublishingMarshaller\",\"getReadonlyMarshaller\",\"getValue\",\"has\",\"ids\",\"lookup\"]")
command[16] E(agoric.bank).help()
history[16] Promise.reject("TypeError: target has no method \"help\", has [\"getAssetSubscription\",\"getPurse\"]")
command[17] E(agoric.chainTimerService).help()
history[17] Promise.reject("TypeError: target has no method \"help\", has []")
command[18] E(agoric.faucet).help()
history[18] Promise.reject("TypeError: target has no method \"help\", has [\"tapFaucet\"]")

command[19] E(agoric.agoricNames).lookup('instance', 'VaultFactory').then(i => E(home.zoe).getPublicFacet(i).then(vfp => E(vfp).help()))
history[19] Promise.reject("TypeError: target has no method \"help\", has []")
```

**Comment by erights** [comment]
> ```
> command[7] E(home.zoe).help()
> history[7] Promise.reject("TypeError: target has no method \"help\", has []")
> ```

That is extremely odd. I don't understand how that could happen. Definitely need to investigate. I will leave this bug open. Thanks!

**Comment by michaelfig** [comment]
> That is extremely odd.

Did you consider my reasoning in https://github.com/Agoric/agoric-sdk/issues/6570#issuecomment-1315870433 ?

**Comment by erights** [comment]
> > That is extremely odd.
> 
> Did you consider my reasoning in [#6570 (comment)](https://github.com/Agoric/agoric-sdk/issues/6570#issuecomment-1315870433) ?

I did. That is what https://github.com/endojs/endo/pull/1382 was supposed to fix. https://github.com/endojs/endo/blob/1d142e58bc3a7388cf851fc74892707f73e4cfd1/packages/eventual-send/src/local.js#L59 is where it ascends the prototype chain. Is the algorithm containing this line wrong, or is the problem elsewhere?

**Comment by michaelfig** [comment]
> or is the problem elsewhere?

We're using different code!  That version of `@endo/eventual-send` has not yet been adopted by agoric-sdk.

**Comment by erights** [comment]
Would be fixed by fixing https://github.com/Agoric/agoric-sdk/issues/7090

**Comment by ivanlei** [comment]
This was fixed after #7090 landed

```
command[3]
E(home.zoe).help()
history[3]
Promise.reject("TypeError: target has no method \"help\", has [\"getBrands\",\"getBundleIDFromInstallation\",\"getConfiguration\",\"getFeeIssuer\",\"getInstallation\",\"getInstallationForInstance\",\"getInstance\",\"getInvitationDetails\",\"getInvitationIssuer\",\"getIssuers\",\"getOfferFilter\",\"getProposalShapeForInvitation\",\"getPublicFacet\",\"getTerms\",\"install\",\"installBundleID\",\"offer\",\"startInstance\"]")
```
